motionplan/constraint.go

```diff
@@ -223,6 +224,70 @@ func NewCollisionConstraintFromWorldState(
 	), nil
 }
 
+// NewOccupancyConstraint creates a constraint that checks for internal geometries of a robot are all ecompassed by the
+// given interaction space geometries at a given input state. If all internal geometries are not encompassed by one
+// interaction space, the constraint will return false.
+func NewOccupancyConstraint(
+	frame referenceframe.Frame,
+	goodInput []referenceframe.Input,
+	iSpaceGeoMap map[string]spatial.Geometry,
+) Constraint {
+	constraint := func(cInput *ConstraintInput) (bool, float64) {
+		internal, err := cInput.Frame.Geometries(cInput.StartInput)
+		if err != nil && internal == nil {
+			return true, 0 // No way of knowing the robot is within a given I-Space without geometries
+		}
+
+		// TODO(wspies): This will break if we have more than one interaction space geometry, unless the arm stayed
+		// inside both geometries, such as for overlapping spaces.
+		internalsEncompassed := true
+		for _, internalGeom := range internal.Geometries() {
+			for _, iSpaceGeom := range iSpaceGeoMap {
+				geomInside, err := internalGeom.EncompassedBy(iSpaceGeom)
+				if err != nil {
+					return false, 0 // Bad encompass check here, just return immediately
+				}
+				internalsEncompassed = internalsEncompassed && geomInside
+			}
+		}
+
+		return internalsEncompassed, 0
+	}
+	return constraint
+}
+
+// NewOccupancyConstraintFromWorldState creates an occupancy constraint from a world state, frame system, a model and a
+// set of initial states.
+func NewOccupancyConstraintFromWorldState(
+	frame referenceframe.Frame,
+	fs referenceframe.FrameSystem,
+	worldState *referenceframe.WorldState,
+	observationInput map[string][]referenceframe.Input,
+) (Constraint, error) {
+	worldState, err := worldState.ToWorldFrame(fs, observationInput)
+	if err != nil {
+		return nil, err
+	}
+
+	// Extract inputs corresponding to the frame
+	var goodInputs []referenceframe.Input
+	switch f := frame.(type) {
+	case *solverFrame:
+		goodInputs, err = f.mapToSlice(observationInput)
+	default:
+		goodInputs, err = referenceframe.GetFrameInputs(f, observationInput)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return NewOccupancyConstraint(
+		frame,
+		goodInputs,
+		worldState.InteractionSpaces[0].Geometries(),
```

Comment:
Rather than only taking the first I-space; can we use the geometries from all interaction spaces? As long as a robot is fully encompassed by at least one, this will work, and functionally there's no difference between one I-space with multiple geometries, and multiple I-spaces.